pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -115,53 +118,83 @@ func (w *Watcher) registerPlugin(socketPath string) error {
-	return nil
+
+	return w.invokeRegistrationCallbackAtHandler(ctx, client, infoResp, socketPath)
 }
 
 func (w *Watcher) invokeRegistrationCallbackAtHandler(ctx context.Context, client registerapi.RegistrationClient, infoResp *registerapi.PluginInfo, socketPath string) error {
 	var handlerCbkFn RegisterCallbackFn
 	var ok bool
 	handlerCbkFn, ok = w.handlers[infoResp.Type]
 	if !ok {
+		errStr := fmt.Sprintf("no handler registered for plugin type: %s at socket %s", infoResp.Type, socketPath)
 		if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
 			PluginRegistered: false,
-			Error:            fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath),
+			Error:            errStr,
 		}); err != nil {
-			glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+			return errors.New(errStr + err.Error())
 		}
-		return fmt.Errorf("no handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath)
+		return errors.New(errStr)
 	}
 
 	var versions []string
 	for _, version := range infoResp.SupportedVersions {
 		versions = append(versions, version)
 	}
 	// calls handler callback to verify registration request
-	err, chanForAckOfNotification := handlerCbkFn(infoResp.Name, infoResp.Endpoint, versions, socketPath)
+	chanForAckOfNotification, err := handlerCbkFn(infoResp.Name, infoResp.Endpoint, versions, socketPath)
 	if err != nil {
+		errStr := fmt.Sprintf("plugin registration failed with err: %v", err)
 		if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
 			PluginRegistered: false,
-			Error:            fmt.Sprintf("Plugin registration failed with err: %v", err),
+			Error:            errStr,
 		}); err != nil {
-			glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+			return errors.New(errStr + err.Error())
 		}
-		chanForAckOfNotification <- false
-		return fmt.Errorf("plugin registration failed with err: %v", err)
+		return errors.New(errStr)
 	}
 
 	if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
 		PluginRegistered: true,
 	}); err != nil {
+		chanForAckOfNotification <- false
 		return fmt.Errorf("failed to send registration status at socket %s, err: %v", socketPath, err)
 	}
+
 	chanForAckOfNotification <- true
 	return nil
 }
 
+// Handle filesystem notify event.
+func (w *Watcher) handleFsNotifyEvent(event fsnotify.Event) error {
+	if event.Op&fsnotify.Create != fsnotify.Create {
+		return nil
+	}
+
+	fi, err := os.Stat(event.Name)
+	if err != nil {
+		return fmt.Errorf("stat file %s failed: %v", event.Name, err)
+	}
+
+	if !fi.IsDir() {
+		return w.registerPlugin(event.Name)
+	}
+
+	if filepath.Base(event.Name)[0] == '.' {
+		return nil
+	}
+
+	if err := w.fsWatcher.Add(event.Name); err != nil {
```

Comment:
in `traversePluginDir` the dir is added, so i guess you don't need to add it here.

Reply:
`Walk(dir)` only walk the file/dir under it, dir itself not included.